### PR TITLE
[GSoC'24] CardBrowser: Implementation of Keyboard Shortcuts

### PR DESCRIPTION
**Pull request by SanjaySargam**
## Purpose / Description
Implement keyboard shortcuts to enhance the user experience by providing quick access to commonly used functionalities within the application. This will improve the overall usability, especially for power users who prefer using the keyboard over the mouse.

## Approach
The shortcuts are implemented by adding event listeners for keypress events and mapping them to the respective functions. The approach ensures that the shortcuts are intuitive and do not conflict with existing shortcuts.

| Action | Shortcut in AnkiDroid | Shortcut in Anki |
|--------|--------|--------|
| Add Tags | Ctrl+Shift+A | Ctrl+Shift+A |
| Export selected cards | Ctrl+Shift+E | Ctrl+Shift+E |
| Reset card progress | Ctrl+Alt+N | Ctrl+Alt+N |
| Toggle cards/notes  | Ctrl+Alt+T | Ctrl+Alt+T | 
| Show filter by tags | T | **n/a**  |
| Resposition selected cards | Ctrl+Shift+S | Ctrl+Shift+S |
| Show saved searches | Ctrl+Alt+S | **n/a** |
| Save search | Ctrl+S | **n/a** |
| Search suspended cards | Alt+S | **n/a** |
| Toggle bury | Ctrl+Shift+J | Ctrl+Shift+J |
| Toggle suspended cards | Ctrl+J | Ctrl+J |
| Card info | Ctrl+Shift+I | Ctrl+Shift+I | 
| Change display order | Ctrl+O | **n/a** |
| Mark note | Ctrl+M | **n/a** |
| Undo | Ctrl+Z | Ctrl+Z |
| Select none | Escape | **n/a** |
| Red flag | Ctrl+1 | Ctrl+1 |
| Orange flag | Ctrl+2 | Ctrl+2 |
| Green flag | Ctrl+3 | Ctrl+3 |
| Blue flag | Ctrl+4 | Ctrl+4 |
| Pink flag | Ctrl+5 | Ctrl+5 |
| Turqoise flag | Ctrl+6 | Ctrl+6 |
| Purple flag | Ctrl+7 | Ctrl+7 |

## How Has This Been Tested?
HP Chromebook

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 1]
```diff
@@ -721,7 +831,12 @@ open class CardBrowser :
     private fun openNoteEditorForCurrentlySelectedNote() = launchCatchingTask {
         try {
             // Just select the first one. It doesn't particularly matter if there's a multiselect occurring.
-            openNoteEditorForCard(viewModel.querySelectedCardIdAtPosition(0))
+            val cardId = if (viewModel.isInMultiSelectMode) {
+                viewModel.querySelectedCardIdAtPosition(0)
+            } else {
+                viewModel.getRowAtPosition(0).id
+            }
+            openNoteEditorForCard(cardId)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:839]: This should be extracted into another method -  it does two separate things, and the method name is no longer correct

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 2]
```diff
@@ -1178,6 +1239,87 @@ open class CardBrowser :
         return super.onOptionsItemSelected(item)
     }
 
+    private fun searchForSuspendedCards() {
+        launchCatchingTask { viewModel.searchForSuspendedCards() }
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1244]: Personal preference 

```suggestion
    private fun searchForSuspendedCards() = launchCatchingTask {
        viewModel.searchForSuspendedCards()
    }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 3]
```diff
@@ -692,8 +720,90 @@ open class CardBrowser :
...
+            }
+            KeyEvent.KEYCODE_S -> {
+                if (event.isCtrlPressed && event.isShiftPressed) {
+                    // Shortcut: Ctrl + Shift + S
+                    Timber.i("Ctrl+Shift+S: reposition selected cards")
+                    repositionSelectedCards()
+                } else if (event.isCtrlPressed && event.isAltPressed) {
+                    // Shortcut: Ctrl + Alt + S
+                    Timber.i("Ctrl+Alt+S: show saved searches")
+                    showSavedSearches()
```
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:751]: It shows resposition dialog

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 1 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:839]: I will make a seperate PR. It is not related to this PR

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:751]: My question was to confirm that this was intended

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 4]
```diff
@@ -718,10 +824,19 @@ open class CardBrowser :
...
     }
 
+    private suspend fun getCardIdForNoteEditor(): CardId {
+        // Just select the first one. It doesn't particularly matter if there's a multiselect occurring.
+        return if (viewModel.isInMultiSelectMode) {
+            viewModel.querySelectedCardIdAtPosition(0)
+        } else {
+            viewModel.getRowAtPosition(0).id
+        }
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:834]: Flagging this for reviewer 2

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 5]
```diff
@@ -634,23 +634,46 @@ open class CardBrowser :
...
+                if (event.isCtrlPressed && event.isShiftPressed) {
+                    Timber.i("Ctrl+Shift+E: Export selected cards")
+                    exportSelected()
+                } else if (event.isCtrlPressed) {
                     Timber.i("Ctrl+E: Add Note")
                     launchCatchingTask { addNoteFromCardBrowser() }
-                    return true
+                } else {
+                    Timber.i("E: Open NoteEditor for selected note")
+                    openNoteEditorForCurrentlySelectedNote()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:674]: I'd like a couple of unit tests here defining behaviour:

* Cards mode/notes mode
* What if there's multiple selections
* What if there's no selection, but rows
* What if there's no rows and no selections

**Review by david-allison — CHANGES_REQUESTED** [review]
I don't feel `E` is well-defined:

I think there's two choices:

* Split it out into a separate PR and get this merged quickly
* Spend some more time, and keep this blocked for a bit

**Comment by SanjaySargam** [comment]
> * Split it out into a separate PR and get this merged quickly

I'd prefer this

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 4 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:834]: didn't understand

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt]: It's not intentional

**Comment by david-allison** [comment]
Sounds great

**Review by david-allison — APPROVED** [review]
LGTM, cheers!

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 6]
```diff
@@ -1178,6 +1228,87 @@ open class CardBrowser :
...
+                    prompt = this@CardBrowser.getString(R.string.reposition_card_dialog_message),
+                    digits = 5
+                )
+                setCallbackRunnable(::repositionCardsNoValidation)
+            }
+            showDialogFragment(repositionDialog)
+        }
+    }
+
+    private fun getCardInfo() {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1303]: As far as I can tell, what it does is displaying to the user the informations of the card. If so, I'd go with `displayCardInfo`

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 7]
```diff
@@ -1178,6 +1228,87 @@ open class CardBrowser :
...
+                    savedFilters,
+                    mySearchesDialogListener,
+                    "",
+                    CardBrowserMySearchesDialog.CARD_BROWSER_MY_SEARCHES_TYPE_LIST
+                )
+            )
+        }
+    }
+
+    private fun saveSearch() {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1263]: Please be specific. Call it `openSaveSearchView` or something similar. Because, reading the code, I thought it was directly saving the search, and I was surprised. By testing; I realize it's not the case.

In particular, the fact that there is not even documentation clarifying is an issue

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 8]
```diff
@@ -692,8 +699,105 @@ open class CardBrowser :
...
+                    repositionSelectedCards()
+                } else if (event.isCtrlPressed && event.isAltPressed) {
+                    Timber.i("Ctrl+Alt+S: Show saved searches")
+                    showSavedSearches()
+                } else if (event.isCtrlPressed) {
+                    Timber.i("Ctrl+S: Save search")
+                    saveSearch()
+                } else if (event.isAltPressed) {
+                    Timber.i("Alt+S: Show suspended cards")
+                    searchForSuspendedCards()
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:731]: If you look at the code in the view model, you'll see it's only expected to be used if the user has no selection.
Indeed, in this case, the action item is not even displayed to the user.

So, there are two problems.
First, you return "true" even if nothing occurred. Ideally, we should return false, so that any system that check whether the shortcut was dealt with does not inform the user that this was a success.

More generally, for each shortcut, it would be nice to document if it's expected to occur when there is a selection, when there is no selection, or whether it does not matter

This is a case where I expect that having `searchForSuspendedCards` return a bool would make sense to me, and then we could directly return the bool

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 9]
```diff
@@ -1178,6 +1228,87 @@ open class CardBrowser :
         return super.onOptionsItemSelected(item)
     }
 
+    private fun searchForSuspendedCards() = launchCatchingTask {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1231]: I'd appreciate if you could document that this is only "replacing the current search by a search for all suspended cards". Intuitively, I would have thought it'd restrict the current search by adding is:suspended, as you can do in anki browser. It's nice to know that it's far more limited than that.

Actually, you could add the documentation in the implementation in viewModel, and just put a `@seeCardBrowserViewModel.searchForSuspendedCards`

**Review by Arthur-Milchior — APPROVED** [review]
I'd be fine merging as is. Even if honestly I'd prefer better name and documentation. And since I expect it's not blocking anything, I won't press "merge"

**Comment by david-allison** [comment]
@Arthur-Milchior @SanjaySargam As far as I'm aware, this is waiting for Sanjay's blessing to merge [review comments are optional].

Should this be merged now, or wait for another round of reviews?

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 10]
```diff
@@ -634,19 +634,26 @@ open class CardBrowser :
         viewModel.setDeckId(deckId)
     }
 
-    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
+    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
         // NOTE: These are all active when typing in the search - doesn't matter as all need CTRL
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:638]: I admit I had a hard time understanding this comment. In particular, I couldn't imagine that if it states that they are "all" active, it implies that it would be possible to have only some of them to be active.

I'd rephrase it as:
> This method is call even when the user is typing a the search text field. So we must ensure that all shortcuts uses a modifier. A shortcut without modifier would be triggered while the user types, which is not what we want.

**Comment by david-allison** [comment]
@Arthur-Milchior Still fine to merge?

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 8 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:736]: I don't understand why this seems to have been mostly ignored.
I appreciate you didn't resolve it. Still, seems strange that you ignored most of it.

For example, for `searcForSuspendedCards`, you added documentation, which I really like. But you omitted what I was asking for in the first place. Documenting that this method does nothing if the selection is not empty, and, in this case, returning `false`

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 11]
```diff
@@ -634,19 +634,28 @@ open class CardBrowser :
         viewModel.setDeckId(deckId)
     }
 
-    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
-        // NOTE: These are all active when typing in the search - doesn't matter as all need CTRL
+    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
+        // This method is call even when the user is typing a the search text field.
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:638]: I'm very sorry for my broken English. I realize I wrote the typos and you just copy-pasted them.
Out of curiosity, did you copy paste without reading. Or had you not realized that what I wrote was wrong.

"is called" "in the search text field"

**Review by Arthur-Milchior — COMMENTED** [review]
@david-allison I admit I'm surprised one comment remained neither answered nor followed.
I'd be fine merging and correcting later if needed. However, as it's not blocking anything, I won't press "merge" until either @SanjaySargam tells us to merge with open comment or follow them.

Also, there are typos that were my mistake, that ended up in the PR

**Comment by david-allison** [comment]
@Arthur-Milchior with apologies, I'm not following, could you link the discussion which requires my input. 

I'm still happy to see this merged

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 11 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:638]: Sorry, I just copy-pasted. I'll make sure to pay more attention next time

**Comment by Arthur-Milchior** [comment]
@david-allison My excuses for being unclear.
You asked whether we could merge. I was trying to answer to you. I was not willing to merge immediately due to one comment which was not resolved.
I was not expecting any answer from you, it was not a question but an answer

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 12]
```diff
@@ -1178,6 +1233,97 @@ open class CardBrowser :
         return super.onOptionsItemSelected(item)
     }
 
+    /**
+     * Returns false if selected rows is not empty else true
+     * @see CardBrowserViewModel.searchForSuspendedCards
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1238]: I'm sorry for being so unclear.
ViewModel's `searchForSuspendedCards` already check whether there is any selected rows. So it was not very useful to do the check here.
Instead, I thought that this function would just be
`private fun searchForSuspendedCards = viewModel.searchForSuspendedCards()` (or actually, you could just remove this method and directly call `viewModel.searchForSuspendedCards()` everywhere, it'll make the code slightly shorter.

The ViewModel's function would do `return false` instead of an early return. And it would return `true` if there are selected rows.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 13]
```diff
@@ -1178,6 +1233,97 @@ open class CardBrowser :
         return super.onOptionsItemSelected(item)
     }
 
+    /**
+     * Returns false if selected rows is not empty else true
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1237]: I think you missed the most important point in this comment. First, that the search is only updated if there is currently no selection.
Secondly, I'd prefer that you explain why this return true or false. That is, you return whether there was any update.

Here is the comment you'd put in the ViewModel.
```
/**
* If there any any selected card, do nothing. 
* Otherwise, searches for all suspended cards and replaces the current search results with these suspended cards.
* @returns whether the search was updated.
*/
```

File: AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt [hunk 14]
```diff
@@ -656,6 +656,9 @@ class CardBrowserViewModel(
         setFilterQuery("tag:marked")
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt:656]: When I wrote "for each shortcut", I meant that you were supposed to look at the implementation of each shortcut.
`searchForMarkedNoted` also do an early return if there is a selection. And this seems to be the same case in a lot of function above and below. So it'd be nice to:
* document that all those methods do nothing if there is a selection (which will probably be the case often when you use the shortcut
* ensure they return a boolean that indicates whether anything was done
* that all method that call them return this boolean, so that the information is propagated to the shortcut function

**Review by Arthur-Milchior — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 12 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1238]: > viewModel.searchForSuspendedCards()

This is suspend function. We can't directly `return viewModel.searchForSuspendedCards()`

Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1238]: Instead should I do like this:
```kotlin
    /**
     * If there any any selected card, do nothing.
     * Otherwise, searches for all suspended cards and replaces the current search results with these suspended cards.
     * @returns whether the search was updated.
     */
    suspend fun searchForSuspendedCards() {
        setFilterQuery("is:suspended")
    }
```

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1256]: I'd like @david-allison opinion on it.
He created the ViewModel, and maybe there are constraints I don't understand.

I admit that, personally, what I would do is:
In the view model
```kotlin
fun searchForSuspendedCards() = setFilterQuery("is:suspended")

/**
Asynchronously edit the search query and list of cards if nothing is selected
@return whether the search update is queued for launch
*/
fun setFilterQuery(filterQuery: String) : False {
  // Do not change the search while some cards are selected
  if (hasSelectedAnyRows()) return False
  launchCatchingTask { 
    // Abort the update if the user selected cards in the meantime
    if (hasSelectedAnyRows()) return
    this.flowOfFilterQuery.emit(filterQuery)
    launchSearchForCards(filterQuery)
  }
  return True
}
```
And then you directly use `viewModel.searchForSuspendedCards()` to deal with both the menu entry and the shortcut.

Of course, that means you must do similar changes to the three methods that use `setFilterQuery()`. Some of them do not have an early return. I believe that adding an early return to them will be a good thing; after all, there is no reason to lose the current search selection when you change the search for tag or flag.



If @david-allison disagrees with my suggestion, then I guess we'll use your suggestion. This sounds like code duplication, but it may be fine

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1256]: David Mentionned I'm overcomplicating things and anyway this code will be changed.
So, fine, let's return true. I'm not huge huge fan of it, but probably reasonable, especially if nobody complain

File: AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt [hunk 15]
```diff
@@ -614,7 +614,9 @@ class CardBrowserViewModelTest : JvmTest() {
         addNoteUsingBasicAndReversedModel("hello2", "world")
 
         runViewModelNotesTest {
-            searchForMarkedNotes()
+            if (!hasSelectedAnyRows()) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt:617]: I'm sorry, but I don't understand why you add the line in the test.
Normally, while running the test, you know whether you expect cards to be selected. And act accordingly.
If you're not 100% , you can also add a test/expectation that the card is selected/not selected.

**Review by Arthur-Milchior — COMMENTED** [review]
One last question regarding tests.

For the Boolean question, I let you decide whether you want to undo it, or not.

File: AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt (hunk 15 above)
Comment (SanjaySargam) [reply AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt:617]: Because I removed this check from `searchForMarkedNotes()`
I'll revert this

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 12 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1262]: For posterity:

> 1. You're overcomplicating things. Return `true`
>     * You handled the keypress and don't want it to bubble. The fact that handling it had no effect is irrelevant
> 2. I have a pull request which removes this functionality, so it's not worth putting in significant effort
>     * Discussed with Sanjay: https://discord.com/channels/368267295601983490/882228005504696350/1268216267601612912
>    * Implemented: [#16859: feat(card-browser): Chip Based Filtering [Tags, Flags & State]](<https://github.com/ankidroid/Anki-Android/issues/16859>)

**Comment by mikehardy** [comment]
(just closing/reopening as I changed the set of github actions that run / are required, this re-triggers CI with the new set)

**Comment by mikehardy** [comment]
@Arthur-Milchior you had this set to auto-merge but...that sat there for 12 hours and did not merge? I perturbed it with my close/reopen for CI required check changes just now. If you intended to merge this I believe you will have to try again?

**Comment by david-allison** [comment]
@Arthur-Milchior you need to approve this

**Review by Arthur-Milchior — APPROVED** [review]
